### PR TITLE
csr <-> csc, csc <-> csc, bsr <-> bsc, bsc <-> bsc, bsr <-> bsr conversions

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. Required to enable a wider selection of sparse formats for `nn.functional.linear`.

### CODE REVIEW

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -992,59 +992,73 @@ Tensor dense_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
...
+      new_values.device());
+}
+
 Tensor sparse_compressed_to_sparse_csr(const Tensor& self) {
   if (self.layout() == kSparseCsc) {
     TORCH_CHECK(
         self.dim() == 2,
         "Expected self to be of dimension 2, but got ",
         self.dim(),
         ".");
```
Comment (amjames) [inline aten/src/ATen/native/TensorConversions.cpp:1057]: It should be possible to support batch dims. A similar approach to what was done to avoid multiple kernel invocations in the `dense->sparse_compressed` conversions could be used. Fold the batch dims into the compressed axis, apply the transformation you have added above to the now 2d CSC matrix, and then unfold.

[This transformation](https://github.com/pytorch/pytorch/blob/4db1588ca09b4f6328c5bd98701e9c302cb39800/aten/src/ATen/native/TensorConversions.cpp#L77-L121) can be used to reintroduce batching in the compressed indices. The reverse of this operation is required for the way in.

Values and plan indices need only be reshaped.

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -2605,11 +2605,13 @@ def _to_from_layout(layout_a, layout_b):
                             or layout_b in [torch.sparse_csc, torch.sparse_bsc])
             expect_error = expect_error or (layout_a, layout_b) == (torch.sparse_bsr, torch.sparse_bsr)
             expect_error = expect_error or (layout_a, layout_b) == (torch.sparse_bsr, torch.sparse_csr)
-            # CSC to CSR conversion is supported
-            if layout_a is torch.sparse_csc and layout_b is torch.sparse_csr:
-                expect_error = False
-            # CSC to CSC conversion is supported
-            if layout_a is torch.sparse_csc and layout_b is torch.sparse_csc:
+
```
Comment (amjames) [inline test/test_sparse_csr.py:2608]: If support is added for batch dims, this test would need to be expanded to generate samples with 3+ dims.

**Review by amjames — COMMENTED** [review]
It is possible to extend this beyond 2d without much work, we could leave that for a follow up though as it is not required supporting `F.linear` with sparse weights only.

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/TensorConversions.cpp:1057]: I am working on it right now.

File: test/test_sparse_csr.py (hunk 2 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:2608]: On it.

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (amjames) [reply aten/src/ATen/native/TensorConversions.cpp:1057]: I don't think that would make the PR too large, @cpuhrsch any opinion on this?

Comment (nikitaved) [reply aten/src/ATen/native/TensorConversions.cpp:1057]: It is a bit tricky as you have to convert to a 1d compressed, then to coo, then you have to modify indices to perform a "transposition" and then reverse it all together. It seems to be more involved as far as I can see it now...

Comment (nikitaved) [reply aten/src/ATen/native/TensorConversions.cpp]: @amjames , I have added batch dims support.

**Comment by nikitaved** [comment]
@amjames , I have done the extension. Could you please have another look?

File: aten/src/ATen/native/TensorConversions.cpp [hunk 3]
```diff
@@ -1381,60 +1577,73 @@ Tensor _csr_to_block_csr_cpu(const Tensor& self, IntArrayRef blocksize) {
...
 Tensor sparse_compressed_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
+  if (self.layout() == kSparseBsr) {
+    return sparse_compressed_to_flipped(self, blocksize, "to_sparse_bsr");
+  }
+  if (self.layout() == kSparseBsc) {
+    return sparse_compressed_clone(self, blocksize, "to_sparse_bsr");
+  }
   AT_ERROR(
-      "Conversion from ", self.layout(), " to SparseBsc is currently not supported.");
+      "sparse_compressed_to_sparse_bsc expected SparseBsr or SparseBsc layout but got ",
```
Comment (amjames) [inline aten/src/ATen/native/TensorConversions.cpp:1631]: We could add support for CSC->BSC using a transpose + conversion of CSR->BSR + transpose back to BSC. It would be 2d only though (since CSR->BSR is 2d only).

`self.transpose(-2, -1).to_sparse_bsr(blocksize).transpose(-2, -1)`

I am fine leaving it out though as it is not particularly useful, this comment is mainly to jog our memory if a request comes in for the functionality.

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -2627,29 +2627,42 @@ def test_compressed_layout_conversions_coverage(self, device, from_layout, to_la
...
-                if (layout_a is not torch.sparse_bsr and layout_b is not torch.sparse_bsr):
-                    self.assertEqual(a.to_dense(), c.to_dense())
+                self.assertEqual(a.to_dense(), c.to_dense())
 
-        _to_from_layout(from_layout, to_layout)
+        batch_dims = [(), (2,), (2, 2), (2, 2, 2)]
+        sparse_dims = (6, 10)
+        for batch_dim in batch_dims:
+            a = make_tensor(batch_dim + sparse_dims, dtype=torch.float, device=device)
+            _to_from_layout(from_layout, to_layout, a)
```
Comment (amjames) [inline test/test_sparse_csr.py:2665]: NIT: Should add at least one test case that exercises the failure condition of the block-size check.

File: test/test_sparse_csr.py [hunk 5]
```diff
@@ -2627,29 +2627,42 @@ def test_compressed_layout_conversions_coverage(self, device, from_layout, to_la
...
+            if (layout_a, layout_b) == (torch.sparse_bsr, torch.sparse_csr):
+                expect_error = True
+            # CSR -> BSR only works for non-batched inputs
+            if (layout_a, layout_b) == (torch.sparse_csr, torch.sparse_bsr):
+                if a.dim() > 2:
+                    expect_error = True
+
             if expect_error:
                 with self.assertRaises(RuntimeError):
                     b = self._convert_to_layout(a, layout_a)
```
Comment (amjames) [inline test/test_sparse_csr.py:2654]: The test is targeting the `layout_a` -> `layout_b` conversion, so the `strided` -> `layout_a` conversion should probably move outside the context manager to ensure the target conversion is the one raising.

**Review by amjames — CHANGES_REQUESTED** [review]
Couple of nits, but requesting one change on the `expect_error` path of the test.

File: aten/src/ATen/native/TensorConversions.cpp [hunk 6]
```diff
@@ -1381,60 +1577,73 @@ Tensor _csr_to_block_csr_cpu(const Tensor& self, IntArrayRef blocksize) {
...
-      c10::kSparseBsr,
-      result_values.device());
+  if (self.layout() == kSparseBsc) {
+    return sparse_compressed_to_flipped(self, blocksize, "to_sparse_bsr");
+  }
+  if (self.layout() == kSparseBsr) {
+    return sparse_compressed_clone(self, blocksize, "to_sparse_bsr");
+  }
+  if (self.layout() == kSparseCsr) {
+    TORCH_CHECK(self.dim() == 2,
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/TensorConversions.cpp:1587]: Right, the goal of this PR is just flipped conversions support.

File: aten/src/ATen/native/TensorConversions.cpp [hunk 7]
```diff
@@ -992,59 +992,255 @@ Tensor dense_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
...
+    const Tensor& self,
+    c10::optional<IntArrayRef> blocksize_opt,
+    const std::string& name) {
+  if (blocksize_opt.has_value()) {
+    const auto blocksize = *blocksize_opt;
+    const auto self_values = self.values();
+    const auto self_blocksize = at::DimVector({self_values.size(-2), self_values.size(-1)});
+    TORCH_CHECK(self_blocksize == blocksize,
+        name, "(): cannot use blocksize (", blocksize[0], ", ", blocksize[1], ") ",
+        "for block compressed input with blocksize (", self_blocksize[0], ", ", self_blocksize[1], ").");
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/TensorConversions.cpp:1005]: I am not sure that the requirement of providing `blocksize` in a block to block conversation is a good design, but it is there, and so is this check.

File: test/test_sparse_csr.py (hunk 5 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:2654]: yeah, that was there before, I did not introduce that part. But I totally agree, will change that.

**Review by amjames — APPROVED** [review]
Awesome, thanks!

**Review by cpuhrsch — APPROVED** [review]
